### PR TITLE
fix(ExportPDF) : #EVAL-436 Fixes some subjects not taking in account subtopic coeffs PDF export when the marks were on a multi-class group.

### PR DESCRIPTION
**Pull request by FlorentMr**
## Describe your changes
Fixes some subjects not taking in account subtopic coeffs PDF export when the marks were on a multi-class group.

## Checklist tests
Suivi Classe -> Moyennes tab -> Check if the pdf has the same averages as in Releve Periodique

## Issue ticket number and link
[EVAL-436](https://jira.support-ent.fr/browse/EVAL-436)

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

**Review by BastienBenard — APPROVED** [review]
Si tu a le temps, ce serait cool de faire des tests sur DefaultSubTopicService.java et DefaultUtilsService.java

**Review by sjollois — APPROVED** [review]

File: src/main/java/fr/openent/competences/service/SubTopicService.java [hunk 1]
```diff
@@ -28,6 +28,14 @@ public interface SubTopicService {
      */
     void getSubtopicServices(String idStructure,String idClasse, Handler<Either<String, JsonArray>> defaultResponseHandler);
 
+    /**
+     * Get subtopic coeffs from an array idsClasse and id Structure
+     * @param idStructure
+     * @param idsClasse
+     * @param handler
+     */
+    void getSubtopicServices(String idStructure, JsonArray idsClasse, Handler<Either<String, JsonArray>> handler);
```
Comment (JinDouang) [inline src/main/java/fr/openent/competences/service/SubTopicService.java:37]: Possibilité de faire en future ?

File: src/main/java/fr/openent/competences/service/UtilsService.java [hunk 2]
```diff
@@ -328,6 +328,14 @@ void getPeriodesClasses (String idEtablissement, JsonArray idClasses,
 
     void getSubTopicCoeff(String idEtablissement, String idClasse, Promise<List<SubTopic>> promise);
 
+    /**
+     * Get subtopic coeffs from an array idsClasse and id Structure
+     * @param idEtablissement
+     * @param idsClasse
+     * @param promise
+     */
+    void getSubTopicCoeff(String idEtablissement, JsonArray idsClasse, Promise<List<SubTopic>> promise);
```
Comment (JinDouang) [inline src/main/java/fr/openent/competences/service/UtilsService.java:337]: Possibilité de faire en future ?

File: src/main/java/fr/openent/competences/service/impl/DefaultSubTopicService.java [hunk 3]
```diff
@@ -67,6 +67,15 @@ public void getSubtopicServices(String idStructure, String idClasse, Handler<Eit
 
     }
 
+    @Override
+    public void getSubtopicServices(String idStructure, JsonArray idsClasse, Handler<Either<String, JsonArray>> handler) {
+        String query = "SELECT  id_subtopic, id_teacher, id_topic, id_group, coefficient::numeric, id_structure " +
+                " From " + this.resourceTable + " WHERE id_structure = ? AND id_group IN " + Sql.listPrepared(idsClasse);
+        JsonArray params = new JsonArray().add(idStructure).addAll(idsClasse);
+        Sql.getInstance().prepared(query,params,validResultHandler(handler));
```
Comment (JinDouang) [inline src/main/java/fr/openent/competences/service/impl/DefaultSubTopicService.java:75]: Possibilité de faire une future avec le model `SubTopic ` ?

File: src/main/java/fr/openent/competences/service/impl/DefaultUtilsService.java [hunk 4]
```diff
@@ -1468,6 +1468,17 @@ public void getSubTopicCoeff(String idEtablissement, String idClasse, Promise<Li
         });
     }
 
+    @Override
+    public void getSubTopicCoeff(String idEtablissement, JsonArray idsClasse, Promise<List<SubTopic>> promise) {
```
Comment (JinDouang) [inline src/main/java/fr/openent/competences/service/impl/DefaultUtilsService.java:1472]: Possibilité de faire une future ?

File: src/main/java/fr/openent/competences/service/impl/DefaultUtilsService.java [hunk 5]
```diff
@@ -1468,6 +1468,17 @@ public void getSubTopicCoeff(String idEtablissement, String idClasse, Promise<Li
...
+        subTopicService.getSubtopicServices(idEtablissement, idsClasse, event -> {
+            if(event.isRight()){
+                setSubtopics(promise, event);
```
Comment (JinDouang) [inline src/main/java/fr/openent/competences/service/impl/DefaultUtilsService.java:1475]: Cette méthode `setSubtopics `pourrait être polymorphé en faisant en sorte que `setSubtopics ` soit une Future

File: src/main/java/fr/openent/competences/service/impl/DefaultSubTopicService.java (hunk 3 above)
Comment (JinDouang) [reply src/main/java/fr/openent/competences/service/impl/DefaultSubTopicService.java:75]:
```suggestion
  Sql.getInstance().prepared(query,params,validResultHandler(event -> {
                        if (event.isRight()) {
                            promise.complete(event.right().getValue())
                        } else {
                            log.error(String.format("[DefaultUtilsService@%s::getSubTopicCoeff] Error during request : %s.",
                                    this.getClass().getSimpleName(), event.left().getValue()));
                            promise.fail(event.left().getValue())
                        }
   }));
```

**Review by JinDouang — CHANGES_REQUESTED** [review]

File: src/main/java/fr/openent/competences/service/impl/DefaultSubTopicService.java [hunk 6]
```diff
@@ -50,29 +56,63 @@ private JsonObject setStatementCoefficient(JsonObject data, String idGroup) {
...
+    @Override
+    public Future<JsonArray> getSubtopicServices(String idStructure, JsonArray idsClasse) {
+        Promise<JsonArray> promise = Promise.promise();
+        String query = "SELECT  id_subtopic, id_teacher, id_topic, id_group, coefficient::numeric, id_structure " +
+                " From " + this.resourceTable + " WHERE id_structure = ? AND id_group IN " + Sql.listPrepared(idsClasse);
+        JsonArray params = new JsonArray().add(idStructure).addAll(idsClasse);
+        return getJsonArrayFuture(promise, query, params);
+    }
 
+    private Future<JsonArray> getJsonArrayFuture(Promise<JsonArray> promise, String query, JsonArray params) {
```
Comment (JinDouang) [inline src/main/java/fr/openent/competences/service/impl/DefaultSubTopicService.java:87]: Pas en paramètre la promise!

**Review by JinDouang — CHANGES_REQUESTED** [review]

**Pull request merged**